### ISSUE TITLE
Linalg Singular matrix in Training

### ISSUE THREAD

**Issue by amine-chaabouni** [yurimjeon1892/EFGH#4]
Hello,

I have downloaded the Rellis_3D dataset and tried to train the model.
In run 1/25, at 58% 4540/7800, i get the following error:

```
_LinAlgError('linalg.inv: (Batch element 0): The diagonal element 2 is zero, the inversion could not be completed because the input matrix is  singular.')
```

Have you experience this befoore or do you know what the problem is? Did the dataset change?
I'm using torch 2.0.0+cu117 instead of torch 1.10.0 since the gpu needs it.
Does that have anything to do with it?

Thank you

**Comment by yurimjeon1892** [comment]
The PyTorch version used in this implementation is 1.x, so there may be issues if you use version 2.x.
Therefore, it is recommended to use a virtual environment such as Docker to run the code using PyTorch version 1.x.

**Comment by amine-chaabouni** [comment]
Thank you for your response.
I used Docker and Conda to have PyTorch 1.10.1 running
I get the following error:
```
'inverse_cpu: (Batch element 0): The diagonal element 3 is zero, the inversion could not be completed because the input matrix is singular.' 
```
It is different, but still the same error.

**Comment by yurimjeon1892** [comment]
Could you provide a detailed explanation of the error? I cannot reproduce the error in my environment. Also, I updated requirement.txt, so please check the package version.

**Comment by amine-chaabouni** [comment]
Hi
I used the new requirements and I still get the same error.
```
[i] Train  2/25:  33%|_____________________________________________________________________                                                                                                                                          | 2582/7800 [1:15:40<1:28:15,  1.01s/batches]
in TRAIN, RuntimeError RuntimeError('inverse_cpu: (Batch element 0): The diagonal element 2 is zero, the inversion could not be completed because the input matrix is singular.')
  File "/root/EFGH/iterater.py", line 38, in iterater
    losses, gt = criterion.compute_loss(pcd, img, calib, A, gt, pred)
  File "/root/EFGH/losses/efghloss.py", line 28, in compute_loss
    floss, gt = self.floss.compute(gt, pred)
  File "/root/EFGH/losses/loss_utils.py", line 80, in compute
    gt_sensor2_T_sensor1_inv = torch.inverse(gt_sensor2_T_sensor1).to(self.device)
```

There is one difference. My GPU is sm_86 (GeForce RTX 3090) and torch==1.10.0 doesn't support that architecture. So I had to install torch using the following command:
`pip install torch==1.10.0+cu111 torchvision==0.11.0+cu111 torchaudio==0.10.0 -f https://download.pytorch.org/whl/torch_stable.html`

You didn't encounter such a problem in your training?

**Comment by yurimjeon1892** [comment]
I have run the code on both the GeForce GTX 1080Ti and GeForce RTX 2080 Ti models, and no such errors were observed. I have updated the Readme file, and the pre-trained model is now available for download. Would you like to give it a try?

**Comment by amine-chaabouni** [comment]
Hi
Thank you for the pre-trained model. I didn't find time to test it before but I did today. However, I'm not sure about how much the model is accurate. Can you have a look at the following results? (using data from Rellis-3D, sequence 00000, data 000103)
This is the point cloud data (from a certain perspective)
![os1_point_cloud_different_PoV](https://user-images.githubusercontent.com/62639815/234308093-0583e316-12a5-4306-8d0f-026519bfe38e.png)
and this is the rgb image
![real_image](https://user-images.githubusercontent.com/62639815/234308105-fd0ca15e-39a9-47c3-9973-00f85b6bde39.png)
As you can see, there is a person in front of the robot and trees on the right, which can be identified on the point cloud as well

However, the registration doesn't register the person in the right spot (I put the saved image on the original one to compare where the registration happened)
![registration](https://user-images.githubusercontent.com/62639815/234308142-1aab0a29-f000-4191-8456-59c042304c33.png)

Also, this result is not always the same from one test to another.

**Comment by yurimjeon1892** [comment]
Regarding accuracy, setting a different initial error range may lead to inaccurate results. 

Additionally, there is a random sampling of points in the code, which can result in different outcomes for the same frame. If you do not want such variability in the results, you can modify the code as follows:

**data_loader/rellis3d_loader.py**
in line **252**, 
`pc_np = pc_np[:, np.random.permutation(pc_np.shape[1])] # comment out this line`

**data_loader/loader_utils.py**
in line **189**..        
 ```
   # Please comment out the code below
    if num_points < pcd.shape[0]:
        sampled_indices1 = \
            np.random.choice(range(pcd.shape[0]),
                                size=num_points, replace=False, p=None)
        pcd_ = pcd[sampled_indices1].T
    else:
        pcd_ = np.zeros(shape=(3, num_points))
        pcd_[:3, :pcd.shape[0]] = pcd[:, :3].T
    
    # and add this
    pcd_ = np.zeros(shape=(3, num_points))
    pcd_[:3, :num_points] = pcd[:num_points, :3].T
```

**Comment by yurimjeon1892** [comment]
Close this issue.

**Comment by negiabhinit** [comment]
I tried doing this Still couldn't make much sense of the results!
Can you help?
![00000_001720_001727_pred_depth_EFH](https://github.com/yurimjeon1892/EFGH/assets/38641617/2ff6db34-591a-47b3-98ee-8789677fb534)

**Comment by yurimjeon1892** [comment]
@negiabhinit Can you open new issue related to this? I'll check it.
